### PR TITLE
Flatcar improvements

### PR DESCRIPTION
**Pull request by FlorinPeter**
**What this PR does / why we need it**:
The PR improves the Flatcar implementation to close the drift between other providers
- remove containerized kubelet
- removed docker service dependency
- use the same containerd config and kubelet.service as other providers
- allow to use the pause image as the sandbox image in containerd to gain more control over the used image

**Special notes for your reviewer**:
This PR was tested with Flatcar 2905.2.3 (Oklo)

**Optional Release Note**:

```release-note
Flatcar improvements
Containerd config improvement
```

### CODE REVIEW

**Comment by FlorinPeter** [comment]
@moadqassem I rebased my Flatcar changes and closed the PR https://github.com/kubermatic/machine-controller/pull/1068 as we discussed today in favor of the upcoming OSM integration. 
Would be nice to get this PR merged to keep the gap between our fork and the upstream as minimal as possible.

File: pkg/userdata/flatcar/provider.go
Comment (kron4eg) [inline pkg/userdata/flatcar/provider.go:268]: we need this file

Comment (FlorinPeter) [reply pkg/userdata/flatcar/provider.go:268]: this is done via the kubeletFlags macro.
You can see it from the expected config here https://github.com/kubermatic/machine-controller/pull/1067/files/a173b5a6aedbf5fbec5461c77ac6077f563db568#diff-1f679fb68a00fca3c52634cc0620d2dc5ee938e7e821bc6790f1233b8acf082cL99-R101

Comment (kron4eg) [reply pkg/userdata/flatcar/provider.go:268]: kubeletFlags uses the ENV that should be set by `EnvironmentFile=/etc/kubernetes/nodeip.conf`, but since it's removed, there is no `$KUBELET_NODE_IP` ENV.

Comment (FlorinPeter) [reply pkg/userdata/flatcar/provider.go:268]: @kron4eg got your point, will check tomorrow why it was left behind, thx for the finding

**Review by kron4eg — APPROVED** [review]

**Comment by kron4eg** [comment]
```
 pkg/userdata/flatcar/provider.go:681: File is not `gofmt`-ed with `-s` (gofmt) 
```

**Comment by FlorinPeter** [comment]
@kron4eg fixed the fmt and a typo on the drop-in, this branch was deployed on a test cluster without any issues

**Review by kron4eg — APPROVED** [review]

**Comment by FlorinPeter** [comment]
@moadqassem something wrong with the PR or why did you put it on hold?

**Comment by moadqassem** [comment]
@FlorinPeter atm vSphere tests are failing due to some internal issues, which means the test will never be green, and since it has LGTM it will keep running the jobs which spam vCenter with no success. We are working on so hopefully will unhold it soon.

**Comment by FlorinPeter** [comment]
@moadqassem thx for the information

File: pkg/userdata/flatcar/provider.go
Comment (moadqassem) [inline pkg/userdata/flatcar/provider.go:215]: Where are you getting this file from? it doesn't seem that you create it anywhere

File: pkg/userdata/flatcar/provider.go
Comment (moadqassem) [inline pkg/userdata/flatcar/provider.go:217]: Since the `/etc/containerd/config.toml` isn't actually created anywhere, this won't start

File: pkg/userdata/flatcar/provider.go
Comment (moadqassem) [inline pkg/userdata/flatcar/provider.go:128]: It seems that this value here isn't loaded correctly

**Review by moadqassem — CHANGES_REQUESTED** [review]
@FlorinPeter Well, this won't work, the reason behind that is, for any k8s version less than 1.22.x we actually deploy docker by default not containerd, this means the systemd services (such as kubelet) assumes that the cri running on the worker node is docker not containerd thus non of those services will start. Maybe I can open another PR which will fox this issue for you but I need to validate couple of things before doing that

**Comment by moadqassem** [comment]
FYI the problems that you had with vCenter other than Flatcar has been resolved. However I am gonna remove the LGTM label since this won't work without couple of changes

**Comment by FlorinPeter** [comment]
> @FlorinPeter Well, this won't work, the reason behind that is, for any k8s version less than 1.22.x we actually deploy docker by default not containerd, this means the systemd services (such as kubelet) assumes that the cri running on the worker node is docker not containerd thus non of those services will start. Maybe I can open another PR which will fix this issue for you but I need to validate couple of things before doing that 

@moadqassem not sure if I understand it correctly. I deployed this change with Kubeone v1.3 and K8s v1.20 without any issue.

**Comment by moadqassem** [comment]
> > @FlorinPeter Well, this won't work, the reason behind that is, for any k8s version less than 1.22.x we actually deploy docker by default not containerd, this means the systemd services (such as kubelet) assumes that the cri running on the worker node is docker not containerd thus non of those services will start. Maybe I can open another PR which will fix this issue for you but I need to validate couple of things before doing that
> 
> @moadqassem not sure if I understand it correctly. I deployed this change with Kubeone v1.3 and K8s v1.20 without any issue.

Yeah sure and I can understand that. K1 deployed containerd as a default cri on all k8s versions unlike machine controller default options. Try change this line here from `docker` to `containerd` and see what you will get: 
https://github.com/kubermatic/machine-controller/blob/master/cmd/machine-controller/main.go#L167

**Comment by FlorinPeter** [comment]
> > > @FlorinPeter Well, this won't work, the reason behind that is, for any k8s version less than 1.22.x we actually deploy docker by default not containerd, this means the systemd services (such as kubelet) assumes that the cri running on the worker node is docker not containerd thus non of those services will start. Maybe I can open another PR which will fix this issue for you but I need to validate couple of things before doing that
> 
> > 
> 
> > @moadqassem not sure if I understand it correctly. I deployed this change with Kubeone v1.3 and K8s v1.20 without any issue.
> 
> 
> 
> Yeah sure and I can understand that. K1 deploys containerd as a default cri on all k8s versions unlike machine controller default options. Try change this line here from `docker` to `containerd` and see what you will get: 
> 
> https://github.com/kubermatic/machine-controller/blob/master/cmd/machine-controller/main.go#L167

Ok now I got the issue you are referring to. Will try to see what will happen if I change it to docker to get a better understanding of the implications.

**Review by moadqassem — COMMENTED** [review]
@FlorinPeter So I have opened a debugging PR based on your changes https://github.com/kubermatic/machine-controller/pull/1092. As you can see, this change https://github.com/kubermatic/machine-controller/pull/1092/files#diff-957f1608260802dae55acd0fed6cbd9009441b087b896d045928c0d0171e6b0fR167 would solve all the issues that you have in the flatcar tests as I suggested earlier. Please include that change in your PR it all tests will pass.

**Comment by FlorinPeter** [comment]
@moadqassem but this will change the default to containerd and will break users that are using docker as default, or? 
I will try to check it tomorrow.

**Comment by moadqassem** [comment]
> @moadqassem but this will change the default to containerd and will break users that are using docker as default, or?
> I will try to check it tomorrow.

How is that a breaking behaviour? In order to use containerd in existing machines, you need to rotate your cluster in the first place, or create a new one. The part where we specify what cri we are gonna use is only at the beginning of machine provisioning. Or what do you exactly mean ?

**Comment by kron4eg** [comment]
I've rebased this PR and fixed issues mentioned by @moadqassem (albeit without changing flags defaults). So this PR is superseded by #1109.

Thanks again @FlorinPeter!

**Comment by FlorinPeter** [comment]
@moadqassem @kron4eg thx a lot, was on sick leave last week and was not able to handle it.
I appreciate your efforts!

**Pull request closed without merging**
